### PR TITLE
Add documentation for connecting to Azure SQL instances with managed identities

### PR DESCRIPTION
**Pull request by lengau**
See issue #6031: https://github.com/sqlalchemy/sqlalchemy/issues/6031

### Description
This adds a subsection to the MSSQL dialect documentation about using Azure credentials to connect to servers.

### Checklist

This pull request is:

- [x] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by zzzeek** [comment]
as mentioned in #6031 this seems overly complicated and also specific to pyodbc, is this documented on pyodbc's website?     for SQLAlhchemy, special arguments like these are just a passthrough for [connect_args](https://docs.sqlalchemy.org/en/13/core/engines.html#use-the-connect-args-dictionary-parameter) and as this guidance seems extremely specific to pyodbc it seems more of a fit for pyodbc's documentation rather than that of SQLAlchemy.

**Comment by lengau** [comment]
In pyodbc itself, you can just say `pyodbc.connect(..., attrs_before={1256: bytes(token_struct})`, as laid out in https://github.com/mkleehammer/pyodbc/issues/228

However, because a sqlalchemy engine can create connections indefinitely into the future, it's necessary to provide a valid token to pyodbc each time we create a connection. Otherwise, the new connection will fail once the initial token has expired. (I was initially using a modified version of the code in the above link, but once my code started running for more than 15 minutes connections started failing.)

Pyodbc doesn't support providing a callable for the value in `attrs_before` so this is the simplest way I've been able to create that solves the problem of long-running engines.

File: lib/sqlalchemy/dialects/mssql/pyodbc.py [hunk 1]
```diff
@@ -79,6 +79,49 @@
...
+Microsoft's Azure SQL allows the use of managed identities to connect to
+databases. This requires creating a credential object using the `azure-identity`
+library. More information about the authentication step can be found in 
+`Microsoft's documentation
+<https://docs.microsoft.com/en-us/azure/developer/python/azure-sdk-authenticate?tabs=bash>`_.
+
+After getting an engine, the credentials need to be sent to ``pyodbc.connect``
+each time a connection is requested. One way to do this is to set up an event
+listener on the engine that adds the credential token to the dialect's connect
+call. The following code snippet allows this:
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/mssql/pyodbc.py:95]: this should end with two colon, so that it's formatted as code. 
tldr `this:` -> `this::`

Comment (lengau) [reply lib/sqlalchemy/dialects/mssql/pyodbc.py]: Thanks for pointing that out!

File: lib/sqlalchemy/dialects/mssql/pyodbc.py [hunk 2]
```diff
@@ -79,6 +79,49 @@
...
+    def inject_azure_credential(credential, engine, token_url='https://database.windows.net'):
+        @event.listens_for(engine, 'do_connect')
+        def do_connect(dialect, conn_rec, cargs, cparams):
+            token = credential.get_token(token_url).token.encode('utf-16-le')
+            token_struct = struct.pack(f'=I{len(token)}s', len(token), token)
+            attrs_before = cparams.setdefault('attrs_before', {})
+            attrs_before[SQL_COPT_SS_ACCESS_TOKEN] = bytes(token_struct)
+            return dialect.connect(*cargs, **cparams)
+            
+This can then be used as follows:
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/mssql/pyodbc.py]: also here `::`

Comment (lengau) [reply lib/sqlalchemy/dialects/mssql/pyodbc.py]: OK this time I actually properly checked the build and made sure it looks right in the browser :-)

**Comment by zzzeek** [comment]
this looks really nice thanks for doing this.  i will pull this through the gerrit process to make sure pep8 still passes and also to backport to 1.3.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision d4bc2a67051ba0886e24d1ee93623780171bb88c of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change d4bc2a67051ba0886e24d1ee93623780171bb88c: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

File: lib/sqlalchemy/dialects/mssql/pyodbc.py [hunk 3]
```diff
@@ -70,15 +69,63 @@
...
+        raw_token = azure_credentials.get_token(TOKEN_URL).token.encode("utf-16-le")
+        token_struct = struct.pack(f"<I{len(raw_token)}s", len(raw_token), token)
+        cparams["attrs_before"] = {SQL_COPT_SS_ACCESS_TOKEN: token_struct}
+        
+.. note::
+    Because `Microsoft requires
+    <https://docs.microsoft.com/en-us/sql/connect/odbc/using-azure-active-directory#authenticating-with-an-access-token>`_
+    that a connection string when using an access token must not contain 
+    ``UID``, ``PWD``, ``Authentication`` or ``Trusted_Connection`` keywords,
+    a the connection string must be provided directly via ``odbc_connect``.
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/mssql/pyodbc.py:128]: typo `a the`

Comment (lengau) [reply lib/sqlalchemy/dialects/mssql/pyodbc.py]: Thanks for catching that. Fixed!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 00abfffd198a4c0e81bb5d15219dd4e40670ce94 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 00abfffd198a4c0e81bb5d15219dd4e40670ce94 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

File: lib/sqlalchemy/dialects/mssql/pyodbc.py [hunk 4]
```diff
@@ -70,15 +69,63 @@
...
+    engine = create_engine(URL.create("mssql+pyodbc", query={"odbc_connect": connection_string}))
+    
+    azure_credentials = identity.DefaultAzureCredential()
+    
+    @event.listens_for(engine, "do_connect")
+    def provide_token(dialect, conn_rec, cargs, cparams):
+        raw_token = azure_credentials.get_token(TOKEN_URL).token.encode("utf-16-le")
+        token_struct = struct.pack(f"<I{len(raw_token)}s", len(raw_token), token)
+        cparams["attrs_before"] = {SQL_COPT_SS_ACCESS_TOKEN: token_struct}
+
```
Comment (sqla-tester) [inline lib/sqlalchemy/dialects/mssql/pyodbc.py:122]: **mike bayer** (zzzeek) wrote:

hey there, can you review what I changed here?  since we are modifying the parameters we can more succinctly get rid of this token in the same method.   if you can make sure this works, this is more succinct.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

i made a few changes if you can review thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

OK I'm going to merge this as is, if you have time to verify the change I made is OK or not, we can always modify it after the fact.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632 has been **merged**. Congratulations! :)
